engine/table/src/main/java/io/deephaven/engine/table/impl/select/analyzers/StaticFlattenLayer.java

```diff
@@ -0,0 +1,133 @@
+package io.deephaven.engine.table.impl.select.analyzers;
+
+import io.deephaven.base.log.LogOutput;
+import io.deephaven.base.verify.Assert;
+import io.deephaven.engine.rowset.RowSet;
+import io.deephaven.engine.rowset.RowSetFactory;
+import io.deephaven.engine.rowset.RowSetShiftData;
+import io.deephaven.engine.rowset.TrackingRowSet;
+import io.deephaven.engine.table.ColumnDefinition;
+import io.deephaven.engine.table.ColumnSource;
+import io.deephaven.engine.table.ModifiedColumnSet;
+import io.deephaven.engine.table.TableUpdate;
+import io.deephaven.engine.table.impl.TableUpdateImpl;
+import io.deephaven.engine.table.impl.sources.RedirectedColumnSource;
+import io.deephaven.engine.table.impl.util.RowRedirection;
+import io.deephaven.engine.table.impl.util.WrappedRowSetWritableRowRedirection;
+
+import java.util.BitSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+final public class StaticFlattenLayer extends SelectAndViewAnalyzer {
+    private final SelectAndViewAnalyzer inner;
+    private final TrackingRowSet parentRowSet;
+    private final Map<String, ColumnSource<?>> overriddenColumns;
+
+    StaticFlattenLayer(SelectAndViewAnalyzer inner, TrackingRowSet parentRowSet) {
+        super(inner.getLayerIndex() + 1);
+        this.inner = inner;
+        this.parentRowSet = parentRowSet;
+        final HashSet<String> alreadyFlattenedColumns = new HashSet<>();
+        inner.getNewColumnSources().forEach((name, cs) -> {
+            alreadyFlattenedColumns.add(name);
+        });
+
+        final RowRedirection rowRedirection = new WrappedRowSetWritableRowRedirection(parentRowSet);
+        overriddenColumns = new HashMap<>();
+        inner.getAllColumnSources().forEach((name, cs) -> {
+            if (alreadyFlattenedColumns.contains(name)) {
+                return;
+            }
+
+            overriddenColumns.put(name, new RedirectedColumnSource<>(rowRedirection, cs));
+        });
+    }
+
+    @Override
+    void setBaseBits(BitSet bitset) {
+        inner.setBaseBits(bitset);
+    }
+
+    @Override
+    void populateModifiedColumnSetRecurse(ModifiedColumnSet mcsBuilder, Set<String> remainingDepsToSatisfy) {
+        inner.populateModifiedColumnSetRecurse(mcsBuilder, remainingDepsToSatisfy);
+    }
+
+    @Override
+    Map<String, ColumnSource<?>> getColumnSourcesRecurse(GetMode mode) {
+        final Map<String, ColumnSource<?>> innerColumns = inner.getColumnSourcesRecurse(mode);
+
+        if (overriddenColumns.keySet().stream().noneMatch(innerColumns::containsKey)) {
```

Comment:
How would this be possible, why would the layer exist at all if you had nothing overridden?

Reply:
This can happen when `mode == GetMode.NEW`.